src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,248 @@
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(ForwardCancellationTokenToAsyncMethodsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingCancellationToken, out INamedTypeSymbol? cancellationTokenType))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+                if (!ShouldAnalyze(
+                    invocation,
+                    cancellationTokenType,
+                    out string? cancellationTokenParameterName))
+                {
+                    return;
+                }
+
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(ForwardCancellationTokenToAsyncMethodsRule, cancellationTokenParameterName, invocation.TargetMethod.Name));
+            },
+            OperationKind.Invocation);
+        }
+
+        private static bool ShouldAnalyze(
+            IInvocationOperation invocation,
+            INamedTypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            IMethodSymbol method = invocation.TargetMethod;
+
+            cancellationTokenParameterName = null;
+
+            // Check if the invocation has an optional implicit ct or an overload that takes one ct
+            if (!InvocationIgnoresOptionalCancellationToken(invocation, cancellationTokenType) &&
+                !MethodHasCancellationTokenOverload(method, cancellationTokenType))
+            {
+                return false;
+            }
+
+            // Find the ancestor method that contains this invocation
+            if (!TryGetAncestorDeclaration(invocation, out IMethodSymbol? methodDeclaration))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, methodDeclaration, out cancellationTokenParameterName))
+            {
+                return false;
+            }
+
+            return cancellationTokenParameterName != null;
```

Comment:
Can this ever be null given `VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(..., [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)`?

```suggestion
            return true;
```